clients/src/main/java/org/apache/kafka/common/memory/RecyclingMemoryPool.java

```diff
@@ -0,0 +1,132 @@
+    protected final int bufferCacheCapacity;
+    protected final AtomicInteger numAllocatedCacheableBuffer;
+    protected final ConcurrentLinkedDeque<ByteBuffer> bufferCache;
+    protected volatile Sensor requestSensor;
+
+    public RecyclingMemoryPool(int cacheableBufferSize, int bufferCacheCapacity, Sensor requestSensor) {
+        if (bufferCacheCapacity <= 0 || cacheableBufferSize <= 0) {
+            throw new IllegalArgumentException(String.format("Must provide a positive cacheable buffer size and buffer cache " +
+                    "capacity, provided %d and %d respectively.", cacheableBufferSize, bufferCacheCapacity));
+        }
+        this.bufferCache = new ConcurrentLinkedDeque<>();
+        this.cacheableBufferSizeUpperThreshold = cacheableBufferSize;
+        this.cacheableBufferSizeLowerThreshold = cacheableBufferSize / 2;
+        this.bufferCacheCapacity = bufferCacheCapacity;
+        this.requestSensor = requestSensor;
+        this.numAllocatedCacheableBuffer = new AtomicInteger(0);
+    }
+
+    @Override
+    public ByteBuffer tryAllocate(int sizeBytes) {
+        if (sizeBytes < 1) {
+            throw new IllegalArgumentException("requested size " + sizeBytes + "<=0");
+        }
+
+        ByteBuffer allocated = null;
+        if (sizeBytes > cacheableBufferSizeLowerThreshold  && sizeBytes <= cacheableBufferSizeUpperThreshold) {
+            allocated = maybePopBufferFromCache();
+        }
+        if (allocated == null) {
+            allocated = ByteBuffer.allocate(sizeBytes);
+        }
+        bufferToBeAllocated(allocated);
+        return allocated;
+    }
+
+    /**
+     * If (1) there are recycled buffers in cache or (2) more buffer of size {@link #cacheableBufferSizeUpperThreshold} can be allocated,
+     * (allocate the buffer and) return the buffer to the client.
+     *
+     * @return The available buffer if any.
+     */
+    protected ByteBuffer maybePopBufferFromCache() {
+        ByteBuffer byteBuffer = bufferCache.poll();
+        if (byteBuffer == null) {
+            int allocated = numAllocatedCacheableBuffer.updateAndGet(v -> v > bufferCacheCapacity ? v : v + 1);
+            if (allocated <= bufferCacheCapacity) {
+                byteBuffer = ByteBuffer.allocate(cacheableBufferSizeUpperThreshold);
+                bufferToBeAllocated(byteBuffer);
+            }
+        }
+        return byteBuffer;
+    }
+
+    @Override
+    public void release(ByteBuffer previouslyAllocated) {
+        if (previouslyAllocated == null) {
+            throw new IllegalArgumentException("provided null buffer");
+        }
+        if (previouslyAllocated.capacity() == cacheableBufferSizeUpperThreshold) {
+            maybeRecycleBufferToCache(previouslyAllocated);
+        }
+        bufferToBeReleased(previouslyAllocated);
+    }
+
+    protected void maybeRecycleBufferToCache(ByteBuffer previouslyAllocated) {
+        if (bufferCache.size() < bufferCacheCapacity) {
```

Comment:
size() is an O(n) operation for ConcurrentLinkedDeque. also - there's a race between size() and offer() in this code that can cause the size of the queue to overshoot the goal. 
I think it will be more correct (and cheaper) for the code to track the size of the list using an atomic integer + updateAndGet() operations

Reply:
Why bother with fancy concurrent data structures if additional complexity is needed.  I'm seeing a big gain from just using a synchronized LinkedList or something.

Reply:
Changed to use [LinkedBlockingQueue](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/LinkedBlockingQueue.html) and preallocate buffers when this memory pool is instantiated (to get rid of `AtomicInteger`).

the good thing about this data structure is 
1. we can specify its capacity during instantiation
2. it has non-blocking [offer()](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/LinkedBlockingQueue.html#offer(E)) and [poll()](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/LinkedBlockingQueue.html#poll()) method which does not throw exception when operation fails.
3. It is thread safe and locking on `offer` and `poll` are separated.